docker/Dockerfile

```diff
@@ -121,15 +121,17 @@ RUN --mount=type=cache,target=/root/.cache/pip \
 COPY synapse /synapse/synapse/
 COPY rust /synapse/rust/
 # ... and what we need to `pip install`.
-COPY pyproject.toml README.rst build_rust.py /synapse/
+COPY pyproject.toml README.rst build_rust.py Cargo.toml Cargo.lock /synapse/
 
 # Repeat of earlier build argument declaration, as this is a new build stage.
 ARG TEST_ONLY_IGNORE_POETRY_LOCKFILE
 
 # Install the synapse package itself.
 # If we have populated requirements.txt, we don't install any dependencies
 # as we should already have those from the previous `pip install` step.
-RUN if [ -z "$TEST_ONLY_IGNORE_POETRY_LOCKFILE" ]; then \
+RUN --mount=type=cache,target=/synapse/target,sharing=locked \
+  --mount=type=cache,target=${CARGO_HOME}/registry,sharing=locked \
```

Comment:
Discussed in [`#synapse-dev`](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$ijoRUxrFIclKKH3_SXwv-LyBAa5LaoIrQlQBCkSc-TU?via=matrix.org&via=element.io&via=beeper.com),

> I think the answer is probably? Otherwise concurrent runs will write over each others caches
> though in practice I don't think you'd really build many docker images in parallel
>
> *-- @erikjohnston*